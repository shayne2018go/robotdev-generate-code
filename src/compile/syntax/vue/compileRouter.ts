import createToken from '@/compile/config/createToken';
import { ICodeSchema } from '@/types';
import { Compile } from '@/types/compile/token';
import { ICS_Directory } from '@/types/directory';
import t from '@babel/types';
import generate from '@babel/generator';

function compileRouter(codeSchema: ICodeSchema): { tokens: Compile.Token[] } {
  const routerPath = 'router/routes.ts';

  const { directories = [], pages } = codeSchema;
  // directories
  const routes = pages.map((page) => {
    return getRouteByDirectories(directories, page.id);
  });

  const tokens = [createToken(routerPath, generateRouterToken(routes))] as Compile.Token[];
  return { tokens };
}

interface VueRoute {
  id: string;
  path: string;
  component: string;
  name?: string;
}

function getRouteByDirectories(directories: ICS_Directory[], id: string): VueRoute {
  const dirPath: ICS_Directory[] = [];

  let curId: string | null = id;

  while (curId) {
    const current = directories.find((d) => d.id === curId);
    if (current) {
      dirPath.unshift(current);
      curId = current.parentId;
    } else {
      break;
    }
  }

  // 解析称 vue route 的path
  const path = `/${dirPath.map((p) => p.key).join('/')}`;

  const component = `../pages${path}.vue`;

  return {
    id,
    name: dirPath[dirPath.length - 1].key,
    path,
    component,
  };
}

function generateRouterToken(routes: VueRoute[]) {
  debugger;
  // 模拟
  const statement = t.blockStatement([
    // 1 导入
    // t.importDeclaration([t.importNamespaceSpecifier(t.identifier('aaa'))], t.stringLiteral('ant-design-vue')),
    t.importDeclaration(
      [t.importSpecifier(t.identifier('RouteRecordRaw'), t.identifier('RouteRecordRaw'))],
      t.stringLiteral('vue-router')
    ),
    // 2 创建routes变量
    t.variableDeclaration('const', [
      t.variableDeclarator(
        t.identifier('routes'),
        t.arrayExpression(
          routes.map((r) =>
            t.objectExpression([
              t.objectProperty(t.identifier('path'), t.stringLiteral(r.path)),
              t.objectProperty(
                t.identifier('component'),
                t.arrowFunctionExpression([], t.callExpression(t.import(), [t.stringLiteral(r.component)]))
              ),
            ])
          )
        )
      ),
    ]),
    t.exportNamedDeclaration(null, [t.exportSpecifier(t.identifier('routes'), t.identifier('routes'))]),
  ]);

  // 3 生成
  const { code } = generate(statement);

  return code;
}

export default compileRouter;
