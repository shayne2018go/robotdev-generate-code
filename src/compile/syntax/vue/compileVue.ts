import { ICodeSchema } from '@/types/ICodeSchema';
import { compileTemplate, compileScript, compileStyle } from './sfc';
import compileRouter from './compileRouter';

function compileVue(codeSchema: ICodeSchema) {
  const { tokens } = compileRouter(codeSchema);

  // const template = compileTemplate(codeSchema);
  // const script = compileScript(codeSchema);
  // const style = compileStyle(codeSchema);
  return tokens.concat(tokens);
}

export default compileVue;
