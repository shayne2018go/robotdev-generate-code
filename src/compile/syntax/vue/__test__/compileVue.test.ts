import codeSchema from '@/__test__/__fixture__/CodeSchema';
import { describe, expect, it } from 'vitest';
import compileRouter from '../compileRouter';
import compileVue, { buildGlobalCtx, parsingVueCompileOptions } from '../compileVue';
import vueRouterTokens from './fixture/vue-router-tokens.json';

describe('compile vue', () => {
  // 解析相关依赖协议
  const vueCompileOptions = parsingVueCompileOptions(codeSchema);

  // 构建全局上下文
  const vueGlobalCtx = buildGlobalCtx(vueCompileOptions);

  it('compileVue', () => {
    const { tokens } = compileVue(codeSchema);
  });

  it('compileRouter', () => {
    debugger;
    const { tokens } = compileRouter(codeSchema, vueGlobalCtx);

    debugger;
    expect(tokens).toMatchSnapshot();
    expect(tokens).toMatchObject(vueRouterTokens);
  });
});
