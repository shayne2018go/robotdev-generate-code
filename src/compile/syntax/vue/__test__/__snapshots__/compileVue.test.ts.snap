// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compile vue > router 1`] = `
[
  {
    "path": "router/routes.ts",
    "token": "{
  import { RouteRecordRaw } from \\"vue-router\\";
  const routes = [{
    path: \\"/shouye\\",
    component: () => import(\\"../pages/shouye.vue\\")
  }, {
    path: \\"/404\\",
    component: () => import(\\"../pages/404.vue\\")
  }, {
    path: \\"/502\\",
    component: () => import(\\"../pages/502.vue\\")
  }, {
    path: \\"/renyuanluruye\\",
    component: () => import(\\"../pages/renyuanluruye.vue\\")
  }, {
    path: \\"/renyuanliebiao\\",
    component: () => import(\\"../pages/renyuanliebiao.vue\\")
  }];
  export { routes };
}",
  },
]
`;
